buttercup/cogs/heatmap.py

```diff
@@ -0,0 +1,175 @@
+    )
+
+    # TODO: Make this dependant on the user's timezone set in the display name
+    timezone = "UTC" if utc_offset == 0 else f"UTC{utc_offset:+d}"
+
+    plt.title(i18n["heatmap"]["plot_title"].format(username))
+    plt.xlabel(i18n["heatmap"]["plot_xlabel"].format(timezone))
+    plt.ylabel(i18n["heatmap"]["plot_ylabel"])
+
+    fig.tight_layout()
+    heatmap_table = io.BytesIO()
+    plt.savefig(heatmap_table, format="png")
+    heatmap_table.seek(0)
+    plt.clf()
+
+    return File(heatmap_table, "heatmap_table.png")
+
+
+def adjust_timezone(heatmap: pd.DataFrame, utc_offset: int) -> pd.DataFrame:
+    """Adjust the heatmap to the given UTC offset."""
+    result = heatmap
+    forward = utc_offset > 0
+
+    # TODO: Make this more efficient
+    # Suppose we have a data frame like this:
+    #     A  B  C
+    # D   1  2  3
+    # E   4  5  6
+    # F   7  8  9
+    # With each iteration, we shift like this:
+    #     A  B  C
+    # D   9  1  2
+    # E   3  4  5
+    # F   6  7  8
+    # Assuming we have a positive UTC offset.
+    for _ in range(0, abs(utc_offset)):
+        periods = 1 if forward else -1
+        wrap_column = result[23] if forward else result[0]
+        wrap_value = wrap_column[7] if forward else wrap_column[1]
+        wrap_column = wrap_column.shift(periods=periods)
+        if forward:
+            wrap_column[1] = wrap_value
+        else:
+            wrap_column[7] = wrap_value
+        result = result.shift(periods=periods, axis="columns")
+        if forward:
+            result[0] = wrap_column
+        else:
+            result[23] = wrap_column
+
+    return result
+
+
+class Heatmap(Cog):
+    def __init__(self, bot: ButtercupBot, blossom_api: BlossomAPI) -> None:
+        """Initialize the Heatmap cog."""
+        self.bot = bot
+        self.blossom_api = blossom_api
+
+    @cog_ext.cog_slash(
+        name="heatmap",
+        description="Display the activity heatmap for the given user.",
+        options=[
+            create_option(
+                name="username",
+                description="The user to get the heatmap for.",
+                option_type=3,
+                required=False,
+            )
+        ],
+    )
+    async def _heatmap(self, ctx: SlashContext, username: Optional[str] = None) -> None:
+        """Generate a heatmap for the given user."""
+        start = datetime.now()
+        user = username or extract_username(ctx.author.display_name)
+        utc_offset = extract_utc_offset(ctx.author.display_name)
+        msg = await ctx.send(i18n["heatmap"]["getting_heatmap"].format(user))
+
+        response = self.blossom_api.get("volunteer/heatmap/", params={"username": user})
+
+        if response.status_code != 200:
+            await msg.edit(content=i18n["heatmap"]["user_not_found"].format(user))
+            return
+
+        data = response.json()
+
+        day_index = pd.Index(range(1, 8))
+        hour_index = pd.Index(range(0, 24))
+
+        heatmap = (

```

Comment:
Wouldn't it be easier to do the offset on the `data` before changing it into a dataframe? e.g.

```
data = [adjust_with_timezone(hour_data, utc_offset) for hour_data in data]

def adjust_with_timezone(hour_data: Dict, utc_offset: int):
  hour_offset = hour_data["hour"] + utc_offset
  new_hour = hour_offset % 24
  new_day = (hour_data["day"] + hour_offset // 24) % 7
  return {"day": new_day, "hour": new_hour, "value": hour_data["value"]
```

Reply:
Much easier! Great algorithm, pretty much worked out of the box except for some small details!